test/functional/testlib.bash

```diff
@@ -68,20 +68,31 @@ generate_certificate() {
 
 	local key_path=$1
 	local cert_path=$2
+	local config=$3
 
 	# If incorrect number of parameters are received show usage
-	if [ $# -ne 2 ]; then
+	if [ $# -eq 0 ]; then
 		cat <<-EOM
 			Usage:
 			    generate_certificate <key_path> <cert_path>
 			EOM
 		return
 	fi
+	validate_param "$key_path"
+	validate_param "$cert_path"
 
 	# generate self-signed public and private key
-	openssl req -x509 -sha256 -nodes -days 365 -newkey rsa:2048 \
-		-keyout "$key_path" -out "$cert_path" \
-		-subj "/C=US/ST=Oregon/L=Portland/O=Company Name/OU=Org/CN=localhost"
+	if [ -z "$config" ]; then
+		openssl req -x509 -sha256 -nodes -days 365 -newkey rsa:2048 \
+			-keyout "$key_path" -out "$cert_path" \
+			-subj "/C=US/ST=Oregon/L=Portland/O=Company Name/OU=Org/CN=localhost"
+	else
+		validate_item "$config"
+		openssl req -x509 -sha256 -nodes -days 365 -newkey rsa:2048 \
+			-keyout "$key_path" -out "$cert_path" \
+			-subj "/C=US/ST=Oregon/L=Portland/O=Company Name/OU=Org/CN=localhost" \
+			-config "$config"
+	fi
```

Comment:
To avoid duplication, can we move the identical sections of the openssl command outside of the condition statement? I think you can store -

config_cmd=""

if [ -n "$config"]; then
           config_cmd="-config $config"
fi

openssl ......... "$config_cmd"

Reply:
I tried that @jwakre  but it didn't work correctly, so I went with adding the full command.